### PR TITLE
Fixes mapping for C# tuples with more than seven elements

### LINKED ISSUE #982: Mapping to a ValueTuple with 8 or more elements only retrieves some columns
I'm not sure if mapping to ValueTuples is supported 100%, but I noticed that if the tuple has 8 or more elements, only the first 7 are mapped and the rest are filled with default values. My guess is that tuples with more items are defined as an 8-tuple with a TRest parameter that is another tuple.

### PR DESCRIPTION
**Pull request by jnm2**
Fixes #982. I think I made the minimum change possible.

/cc @mgravell

### CODE REVIEW

File: Dapper/SqlMapper.cs [hunk 1]
```diff
@@ -3097,6 +3068,121 @@ private static LocalBuilder GetTempLocal(ILGenerator il, ref Dictionary<Type, Lo
...
+            else
+            {
+                GenerateDeserializerFromMap(type, reader, startBound, length, returnNullIfFirstMissing, il);
+            }
+
+            var funcType = System.Linq.Expressions.Expression.GetFuncType(typeof(IDataReader), returnType);
+            return (Func<IDataReader, object>)dm.CreateDelegate(funcType);
+        }
+
+        private static void GenerateValueTupleDeserializer(Type valueTupleType, IDataReader reader, int startBound, int length, bool returnNullIfFirstMissing, ILGenerator il)
```
Comment (jnm2) [inline Dapper/SqlMapper.cs]: @mgravell (For review, before I forget) The parameter `returnNullIfFirstMissing` is unused. Should I be using it, or can I rule that out and remove the parameter since the deserialization type is a value type?

File: Dapper/SqlMapper.cs [hunk 2]
```diff
@@ -3097,6 +3068,121 @@ private static LocalBuilder GetTempLocal(ILGenerator il, ref Dictionary<Type, Lo
...
+
+            var enumDeclareLocal = -1;
+
+            for (var i = 0; i < languageTupleElementTypes.Count; i++)
+            {
+                var targetType = languageTupleElementTypes[i];
+
+                if (i < length)
+                {
+                    LoadReaderValueOrBranchToDBNullLabel(
```
Comment (jnm2) [inline Dapper/SqlMapper.cs]:
```suggestion
                    LoadReaderValueOrBranchToDbNullLabel(
```

**Comment by jnm2** [comment]
@mgravell Would you be able to set my expectations for an ETA? I hate to bug you.

**Comment by mgravell** [comment]
You're right, I need to find the time. I'll see if I can review on Tuesday
(long family weekend here).

File: Dapper/SqlMapper.cs (hunk 1 above)
Comment (mgravell) [reply Dapper/SqlMapper.cs]: yeah, I think you're fine to remove it... but could always just add a test that adds a `(int,int)?` type and zero rows?

File: Dapper/SqlMapper.cs [hunk 3]
```diff
@@ -3097,6 +3068,121 @@ private static LocalBuilder GetTempLocal(ILGenerator il, ref Dictionary<Type, Lo
...
+        private static void GenerateValueTupleDeserializer(Type valueTupleType, IDataReader reader, int startBound, int length, bool returnNullIfFirstMissing, ILGenerator il)
+        {
+            var currentValueTupleType = valueTupleType;
+
+            var constructors = new List<ConstructorInfo>();
+            var languageTupleElementTypes = new List<Type>();
+
+            while (true)
+            {
+                var arity = int.Parse(currentValueTupleType.Name.Substring("ValueTuple`".Length));
```
Comment (mgravell) [inline Dapper/SqlMapper.cs]: needs to be invariant?

Comment (mgravell) [reply Dapper/SqlMapper.cs]: (meaning: culture)

File: Dapper/SqlMapper.cs [hunk 4]
```diff
@@ -3097,6 +3068,121 @@ private static LocalBuilder GetTempLocal(ILGenerator il, ref Dictionary<Type, Lo
...
+
+                foreach (var field in currentValueTupleType.GetFields(BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly))
+                {
+                    if (field.Name == "Rest")
+                    {
+                        restField = field;
+                    }
+                    else if (field.Name.StartsWith("Item", StringComparison.Ordinal))
+                    {
+                        var elementNumber = int.Parse(field.Name.Substring("Item".Length));
```
Comment (mgravell) [inline Dapper/SqlMapper.cs]: ditto

**Comment by mgravell** [comment]
I've given it a once over - looks kinda great, which is nice; I'm cautious about the `int.Parse` because: culture, so that might need a check; it is also possible that I'm wrong and it works fine already. I want to read it from scratch a second time, and look through the tests - but: looking very promising.

File: Dapper/SqlMapper.cs (hunk 3 above)
Comment (jnm2) [reply Dapper/SqlMapper.cs]: Oops, this is something I keep forgetting to do. Thanks!

**Comment by jnm2** [comment]
Awesome! I'll write that nullable tuple test and fix the culture sensitivity either late tonight or sometime tomorrow and ping you when I'm finished.

After you get a chance to do the final review, I'll be curious if there's a prerelease feed.

**Comment by mgravell** [comment]
Preview feed on myget via appveyor:
https://www.myget.org/feed/stackoverflow/package/nuget/Dapper

File: Dapper/SqlMapper.cs (hunk 1 above)
Comment (jnm2) [reply Dapper/SqlMapper.cs]: @mgravell `connection.QuerySingleOrDefault<(int, int)?>("select null where 0 = 1");` does not result in `returnNullIfFirstMissing` becoming true. Should I be trying something with multi-mapping? (Looks like the parameter was added in https://github.com/StackExchange/Dapper/commit/e11257fdf82a4307523e09859e31da3d495877a6)

Comment (jnm2) [reply Dapper/SqlMapper.cs]: I wrote this test against master to prove that nullable C# tuples have never been supported:

```cs
[Fact]
public void NullableTupleMultiMap()
{
    var outer = connection.Query<Foo, (int, int)?, Foo>(
        "select * from (values (42, null, null), (42, 43, 44)) as Foo (A, B1, B2)",
        (foo, b) =>
        {
            foo.B = b;
            return foo;
        },
        splitOn: "B1");

    Assert.Equal(42, outer.First().A);
    Assert.Null(outer.First().B);

    Assert.Equal(42, outer.Last().A);
    Assert.Equal(43, outer.Last().B.Value.Item1);
    Assert.Equal(44, outer.Last().B.Value.Item2);
}

private sealed class Foo
{
    public int A { get; set; }
    public (int, int)? B { get; set; }
}
```

And therefore `returnNullIfFirstMissing` is not meaningful yet and I can remove it.

**Comment by jnm2** [comment]
@mgravell Ready for the second round!

**Comment by jnm2** [comment]
The force pushes above are [adding `CultureInfo.InvariantCulture`](https://github.com/StackExchange/Dapper/compare/31bf76f1eb4c9b5897ac1b43dd559413d6efe934..93245d2a390aa7c2064f644be7e60305f0011d76) and [removing `returnNullIfFirstMissing`](https://github.com/StackExchange/Dapper/compare/93245d2a390aa7c2064f644be7e60305f0011d76..f61c3ce6ec1dc165fa4c6ff4827c666eae3547d6), addressing both your comments.

Do you think you might have a chance coming up next week to review again? I appreciate the time you've been able to give.

**Comment by jnm2** [comment]
@mgravell The workaround at the moment is to introduce new private throwaway structs, so this keeps coming up. Is this PR a very risky one? Would it help if I added more tests?

**Comment by mgravell** [comment]
let's do this...

**Comment by jnm2** [comment]
We'll start using the MyGet feed right away! Thank you so much! 🎉

**Comment by mgravell** [comment]
No: thank you. It was your contribution, and it is very much appreciated. Delay here was mostly my time juggling. Thanks again.

**Comment by jnm2** [comment]
As an OSS maintainer, I know PRs are demanding. I appreciate you reviewing and getting this through.

File: Dapper/SqlMapper.cs [hunk 5]
```diff
@@ -3400,9 +3391,115 @@ private static LocalBuilder GetTempLocal(ILGenerator il, ref Dictionary<Type, Lo
...
+            }
+        }
+
+        private static void LoadReaderValueOrBranchToDBNullLabel(ILGenerator il, int index, ref int enumDeclareLocal, int? valueCopyLocal, Type colType, Type memberType, out Label isDbNullLabel)
+        {
+            isDbNullLabel = il.DefineLabel();
+            il.Emit(OpCodes.Ldarg_0); // stack is now [...][reader]
+            EmitInt32(il, index); // stack is now [...][reader][index]
+            il.Emit(OpCodes.Dup);// stack is now [...][reader][index][index]
+            il.Emit(OpCodes.Stloc_0);// stack is now [...][reader][index]
```
Comment (jnm2) [inline Dapper/SqlMapper.cs]: This line introduced bug https://github.com/StackExchange/Dapper/issues/1279, fixed in https://github.com/StackExchange/Dapper/pull/1280.
